Nodejs/Product/Nodejs/SharedProject/Wpf/NotificationTextBox.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Windows.Automation.Peers;
+using System.Windows.Automation.Provider;
+using System.Windows.Controls;
+
+namespace Microsoft.NodejsTools.SharedProject.Wpf
+{
+    internal class NotificationTextBox : TextBox
+    {
+        // This control's AutomationPeer is the object that actually raises the UIA Notification event.
+        private NotificationTextBoxAutomationPeer notificationTextBoxAutomationPeer;
+
+        // Assume the UIA Notification event is available until we learn otherwise.
+        // If we learn that the UIA Notification event is not available, no instance
+        // of the NotificationTextBox should attempt to raise it.
+        private static bool notificationEventAvailable = true;
+        public bool NotificationEventAvailable
+        {
+            get
+            {
+                return notificationEventAvailable;
+            }
+            set
+            {
+                notificationEventAvailable = value;
+            }
+        }
+
+        protected override AutomationPeer OnCreateAutomationPeer()
+        {
+            this.notificationTextBoxAutomationPeer = new NotificationTextBoxAutomationPeer(this);
+
+            return this.notificationTextBoxAutomationPeer;
+        }
+
+        public void RaiseNotificationEvent(string notificationText, string notificationGuid)
+        {
+            // Only attempt to raise the event if we already have an AutomationPeer.
+            if (this.notificationTextBoxAutomationPeer != null)
+            {
+                this.notificationTextBoxAutomationPeer.RaiseNotificationEvent(notificationText, notificationGuid);
+            }
+        }
+    }
+
+    internal class NotificationTextBoxAutomationPeer : TextBoxAutomationPeer
+    {
+        private NotificationTextBox notificationTextBox;
+
+        // The UIA Notification event requires the IRawElementProviderSimple
+        // associated with this AutomationPeer.
+        private IRawElementProviderSimple rawElementProviderSimple;
+
+        public NotificationTextBoxAutomationPeer(NotificationTextBox owner) : base(owner)
+        {
+            this.notificationTextBox = owner;
+        }
+
+        public void RaiseNotificationEvent(string notificationText, string notificationGuid)
+        {
+            // If we already know that the UIA Notification event is not available, do not
+            // attempt to raise it.
+            if (this.notificationTextBox.NotificationEventAvailable)
```

Comment:
Why is this stored on the textbox if it is only read and written by the peer?

Reply:
Only for the intention of exposing it on the TextBox element. I'll could move it to the peer if you feel its necessary.

Reply:
Oh, it's used in XAML too?  Good answer.  Just leave it. :)